expression/constant_fold.go

```diff
@@ -78,6 +79,47 @@ func ifNullFoldHandler(expr *ScalarFunction) (Expression, bool) {
 	return expr, isDeferredConst
 }
 
+func caseWhenHandler(expr *ScalarFunction) (Expression, bool) {
+	args, l := expr.GetArgs(), len(expr.GetArgs())
+	var isDeferred, isDeferredConst, hasNonConstCondition bool
+	for i := 0; i < l-1; i += 2 {
+		expr.GetArgs()[i], isDeferred = foldConstant(args[i])
+		isDeferredConst = isDeferredConst || isDeferred
+		if _, isConst := expr.GetArgs()[i].(*Constant); isConst && !hasNonConstCondition {
+			// If the condition is const and true, and the previous conditions
+			// has no expr, then the folded execution body is returned, otherwise
+			// the arguments of the casewhen are folded and replaced.
+			val, isNull, err := args[i].EvalInt(expr.GetCtx(), chunk.Row{})
+			if err != nil {
+				return expr, false
+			}
+			if val != 0 && !isNull {
+				foldedExpr, isDeferred := foldConstant(args[i+1])
+				isDeferredConst = isDeferredConst || isDeferred
+				return BuildCastFunction(expr.GetCtx(), foldedExpr, expr.GetType()), isDeferredConst
+			}
+		} else {
+			hasNonConstCondition = true
+		}
+		expr.GetArgs()[i+1], isDeferred = foldConstant(args[i+1])
+		isDeferredConst = isDeferredConst || isDeferred
+	}
+
+	if l%2 == 1 && !hasNonConstCondition {
```

Comment:
how about:
```go
if l%2 == 0 {
	return expr, isDeferredConst
}
if !hasNonConstCondition {
	// do something
}
// do something
```

Reply:
Good advice, I have modified it. PTAL @zz-jason